CMakeLists.txt

```diff
@@ -650,45 +783,99 @@ foreach(clangLib ${CLANG_LIBRARY_LIST})
         list(APPEND XE_LIBRARY_FULL_PATH_LIST ${${xeLib}Path})
     endforeach()
-    target_link_libraries(${PROJECT_NAME} ${XE_LIBRARY_FULL_PATH_LIST})
+    list(APPEND LINK_LIBRARIES ${XE_LIBRARY_FULL_PATH_LIST})
 endif()
 
 if (NOT WIN32 AND NOT APPLE)
-    target_link_libraries(${PROJECT_NAME} "-Wl,--end-group")
+    list(APPEND LINK_LIBRARIES "-Wl,--end-group")
 endif()
 
 # System libraries, our own and transitive dependencies from LLVM libs.
 if (WIN32)
-    target_link_libraries(${PROJECT_NAME} version.lib shlwapi.lib odbc32.lib odbccp32.lib)
+    list(APPEND LINK_LIBRARIES version.lib shlwapi.lib odbc32.lib odbccp32.lib)
 else()
     if (ISPC_STATIC_LINK)
-        target_link_libraries(${PROJECT_NAME} pthread tinfo.a curses.a)
+        list(APPEND LINK_LIBRARIES pthread tinfo.a curses.a)
     else()
         find_package(Curses REQUIRED)
         if (CURSES_EXTRA_LIBRARY)
             # this contains the tinfo library, if found
-            target_link_libraries(${PROJECT_NAME} ${CURSES_EXTRA_LIBRARY})
+            list(APPEND LINK_LIBRARIES ${CURSES_EXTRA_LIBRARY})
         endif()
         # Recommended way to get Curses library is to use CURSES_LIBRARIES variable
         # https://cmake.org/cmake/help/v3.27/module/FindCurses.html
         # but it contains both libform.so and libcurses.so.
         # We need libcurses.so only, so exclude libform.so from the list.
         list(FILTER CURSES_LIBRARIES EXCLUDE REGEX ".*libform.*")
-        target_link_libraries(${PROJECT_NAME} pthread ${CURSES_LIBRARIES})
+        list(APPEND LINK_LIBRARIES pthread ${CURSES_LIBRARIES})
+    endif()
+endif()
+
+
+# This function configures compile definitions, options and include dirs.
+function(configure_ispc_obj TARGET)
+    target_compile_definitions(${TARGET} PRIVATE ${COMPILE_DEFINITIONS})
+    target_include_directories(${TARGET} PRIVATE ${INCLUDE_DIRECTORIES})
+    target_compile_options(${TARGET} PRIVATE ${COMPILE_OPTIONS})
+
+    # Set hidden visibility for inline functions.  This is needed to be in sync
+    # with LLVM libraries, we do include some template code from LLVM that requires
+    # this options to be in sync with LLVM. Otherwise it causes link warnings.
+    set_target_properties(${TARGET} PROPERTIES VISIBILITY_INLINES_HIDDEN ON)
+    if (NOT MSVC)
+        set_target_properties(${TARGET} PROPERTIES CXX_EXTENSIONS OFF)
     endif()
+endfunction()
+
+# This function configures executable targets ispc-slim and ispc-whole.
+function(configure_ispc TARGET TARGET_SPECIFIC_SOURCES)
+    add_executable(${TARGET} ${TARGET_SPECIFIC_SOURCES})
+    configure_ispc_obj(${TARGET})
+    target_link_options(${TARGET} PRIVATE ${LINK_OPTIONS_PRIVATE})
+    target_link_options(${TARGET} PUBLIC ${LINK_OPTIONS_PUBLIC})
+    target_link_libraries(${TARGET} ${LINK_LIBRARIES} common)
+endfunction()
+
+# Target common contains all sources that are shared between slim and whole binaries.
```

Comment:
An idea to refer to the old binary as "composite".
"slim" could also be named as "lean". Just a suggestion for consideration.

Reply:
I'm OK with "slim" and I like the idea to have "composite" instead of "whole".